File: src/main/java/com/ordermanagement/service/CategoryService.java
package com.ordermanagement.service;

import com.ordermanagement.domain.entity.Category;
import com.ordermanagement.dto.category.CategoryResponse;
import com.ordermanagement.dto.category.CreateCategoryRequest;
import com.ordermanagement.exception.DuplicateResourceException;
import com.ordermanagement.exception.ResourceNotFoundException;
import com.ordermanagement.repository.CategoryRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CategoryService {

    private final CategoryRepository categoryRepository;

    @Transactional
    public CategoryResponse create(CreateCategoryRequest request) {
        if (categoryRepository.existsByName(request.getName())) {
            throw new DuplicateResourceException("Category", "name", request.getName());
        }

        Category category = Category.builder()
                .name(request.getName())
                .description(request.getDescription())
                .active(true)
                .build();

        categoryRepository.save(category);

        return CategoryResponse.fromEntity(category);
    }

    @Transactional(readOnly = true)
    public CategoryResponse findById(Long id) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category", id));

        return CategoryResponse.fromEntity(category);
    }

    @Transactional(readOnly = true)
    public List<CategoryResponse> findAllActive() {
        return categoryRepository.findByActiveTrue().stream()
                .map(CategoryResponse::fromEntity)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<CategoryResponse> findAll() {
        return categoryRepository.findAll().stream()
                .map(CategoryResponse::fromEntity)
                .collect(Collectors.toList());
    }

    @Transactional
    public CategoryResponse update(Long id, CreateCategoryRequest request) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category", id));

        category.setName(request.getName());
        category.setDescription(request.getDescription());

        categoryRepository.save(category);

        return CategoryResponse.fromEntity(category);
    }

    @Transactional
    public void deactivate(Long id) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category", id));

        category.setActive(false);
        categoryRepository.save(category);
    }

    @Transactional(readOnly = true)
    public Category getEntityById(Long id) {
        return categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category", id));
    }
}

File: internal/service/category_service.go
package service

import (
	"errors"

	"github.com/refleeexzz/order-management/internal/apperror"
	"github.com/refleeexzz/order-management/internal/domain"
	"github.com/refleeexzz/order-management/internal/dto"
	"github.com/refleeexzz/order-management/internal/repository"
	"gorm.io/gorm"
)

// CategoryService mirrors service/CategoryService.
type CategoryService struct {
	categories *repository.CategoryRepository
}

func NewCategoryService(categories *repository.CategoryRepository) *CategoryService {
	return &CategoryService{categories: categories}
}

// Create mirrors CategoryService.create: duplicate name (case-sensitive
// exists check) → 409; the new category is active=true.
func (s *CategoryService) Create(req dto.CreateCategoryRequest) (*dto.CategoryResponse, error) {
	exists, err := s.categories.ExistsByName(req.Name)
	if err != nil {
		return nil, err
	}
	if exists {
		return nil, apperror.Duplicate("Category", "name", req.Name)
	}

	category := &domain.Category{
		Name:        req.Name,
		Description: req.Description,
		Active:      true,
	}
	if err := s.categories.Create(category); err != nil {
		return nil, err
	}
	return s.response(category)
}

// FindByID mirrors CategoryService.findById → 404 when absent.
func (s *CategoryService) FindByID(id uint) (*dto.CategoryResponse, error) {
	category, err := s.GetEntityByID(id)
	if err != nil {
		return nil, err
	}
	return s.response(category)
}

// FindAllActive mirrors CategoryService.findAllActive (active only).
func (s *CategoryService) FindAllActive() ([]dto.CategoryResponse, error) {
	categories, err := s.categories.FindAllActive()
	if err != nil {
		return nil, err
	}
	return s.toResponses(categories)
}

// FindAll mirrors CategoryService.findAll (includes inactive — admin only).
func (s *CategoryService) FindAll() ([]dto.CategoryResponse, error) {
	categories, err := s.categories.FindAll()
	if err != nil {
		return nil, err
	}
	return s.toResponses(categories)
}

// Update mirrors CategoryService.update: name+description only, and NO
// duplicate-name re-check (Java parity).
func (s *CategoryService) Update(id uint, req dto.CreateCategoryRequest) (*dto.CategoryResponse, error) {
	category, err := s.GetEntityByID(id)
	if err != nil {
		return nil, err
	}

	category.Name = req.Name
	category.Description = req.Description
	if err := s.categories.Save(category); err != nil {
		return nil, err
	}
	return s.response(category)
}

// Deactivate mirrors CategoryService.deactivate: soft delete (active=false).
func (s *CategoryService) Deactivate(id uint) error {
	category, err := s.GetEntityByID(id)
	if err != nil {
		return err
	}
	category.Active = false
	return s.categories.Save(category)
}

// GetEntityByID mirrors CategoryService.getEntityById — the internal
// accessor ProductService uses to resolve/validate categories
// (404 "Category not found with id: <id>").
func (s *CategoryService) GetEntityByID(id uint) (*domain.Category, error) {
	category, err := s.categories.FindByID(id)
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, apperror.NotFound("Category", int64(id))
	}
	if err != nil {
		return nil, err
	}
	return category, nil
}

func (s *CategoryService) toResponses(categories []domain.Category) ([]dto.CategoryResponse, error) {
	responses := make([]dto.CategoryResponse, 0, len(categories))
	for i := range categories {
		resp, err := s.response(&categories[i])
		if err != nil {
			return nil, err
		}
		responses = append(responses, *resp)
	}
	return responses, nil
}

// response maps a category to its DTO, counting ALL products of the
// category (incl. inactive) like fromEntity does via the lazy products
// collection.
func (s *CategoryService) response(category *domain.Category) (*dto.CategoryResponse, error) {
	count, err := s.categories.CountProducts(category.ID)
	if err != nil {
		return nil, err
	}
	return dto.NewCategoryResponse(category, int(count)), nil
}
